tests/image_test.py

```diff
@@ -751,6 +751,63 @@ def test_resize_and_sample_errors(method, fx_asset):
             getattr(img, method)(height=-5)
 
 
+@mark.parametrize(( 'density', 'expected_size'), [
+    ((72,72), (800, 600)),
+    ((36,36), (400, 300)),
+    ((144,144), (1600, 1200)),
+    ((None,36), (800, 300)),
+    ((36,None), (400, 600)),
+])
+def test_resample(density, expected_size, fx_asset):
+    """Resample (Adjust nuber of pixels at the given density) the image."""
+    xr, yr = density
+    with Image(filename=str(fx_asset.join('beach.jpg'))) as img:
+        img.units = "pixelspercentimeter"
+        assert img.resolution == (72, 72)
+        img.resample(xr, yr)
+        assert img.resolution == (xr == None and 72 or xr, yr == None and 72 or yr)
+        assert img.size == expected_size
+
+
+@mark.parametrize(( 'density', 'expected_size'), [
+    ((96,96), (350, 197)),
+    ((48,48), (175, 99)),
+    ((192,192), (700, 394)),
+    ((None,48), (350, 99)),
+    ((48,None), (175, 197)),
+])
+def test_resample_gif(density, expected_size, fx_asset):
+    """Resample (Adjust nuber of pixels at the given density) the image."""
```

Comment:
This test is fomally correct, but it fails.
GIF don't support "density" as standard. Maybe that setter of Image.resolution should be changed to set resolution for each frame in case of animated gif.